File: kompartment/src/ode/rosenbrock23.js
/**
 * A Rosenbrock (2,3) pair, for stiff systems.
 *
 * The method is the linearly implicit W-method described by L. F. Shampine
 * and M. W. Reichelt, "The MATLAB ODE Suite", SIAM J. Sci. Comput. 18(1) 1-22
 * (1997), section 4: two stages, one Jacobian and one factorisation per step,
 * and an embedded third-order estimate to judge the step by. With
 * d = 1/(2 + √2) and W = I - h·d·J, the stages are
 *
 *     W k1 = f(t, y) + h·d·f_t
 *     W k2 = f(t + h/2, y + h/2·k1) - k1               (then k2 += k1)
 *     y_{n+1} = y + h·k2
 *     W k3 = f(t + h, y_{n+1}) - (6 + √2)(k2 - f(t + h/2, ·)) - 2(k1 - f(t, y)) + h·d·f_t
 *
 * and the error estimate is (h/6)(k1 - 2k2 + k3). It is the solver that
 * matters for a radionuclide chain whose half-lives span many orders of
 * magnitude: an explicit method grinds its step to nothing on the fastest
 * nuclide, and this one does not.
 *
 * The iteration matrix is factorised densely through ./linalg.js, or in CSC
 * through ./sparse.js where a supplied pattern pays for it. Everything around
 * the step -- step-size control, events, output, the constraint -- is the
 * shared driver in ./onestep.js.
 *
 * The published pair has no non-negativity option. This one takes the
 * option's meaning from the methods that have it: the derivative of a state
 * on its bound may not carry it lower, a violation larger than the tolerance
 * fails the error test and cuts the step, and what is left below zero after
 * an accepted step is projected back and counted. A state held at zero keeps
 * delivering whatever its outflows say, so an empty compartment drained at an
 * absolute rate of one for ten years still hands ten units to its neighbour
 * -- the projected system the option asks for. A constraint that takes hold
 * *inside* a run is another matter: a one-step method's stages straddle the
 * kink and disagree by the whole jump, so the step is cut and cut again, and
 * the driver's stall guard is what ends that with an explanation.
 */

import { LU } from './linalg.js';
import { sparseIterationMatrix } from './sparse.js';
import { integrate, SolverError } from './onestep.js';

/** What this tool adds to the shared sparse LU's singular message. */
const SINGULAR_HINT = 'A compartment with no way in and no way out will do this.';

const D = 1 / (2 + Math.SQRT2);
const E32 = 6 + Math.SQRT2;
const EPS = 2 ** -52;
const SQRT_EPS = Math.sqrt(EPS);

/**
 * df/dy by forward differences of the derivative the method integrates, one
 * column per evaluation, or through a sparsity pattern and a colouring of its
 * columns when one was supplied. The increment for a column is sqrt(eps)
 * times the larger of the state and its error threshold, rounded to what the
 * addition actually changed.
 */
function differences(rhs, neq, threshold) {
	const ytry = new Float64Array(neq);
	const ftry = new Float64Array(neq);
	const delta = new Float64Array(neq);
	const increment = (y, j) => {
		let del = SQRT_EPS * Math.max(Math.abs(y[j]), threshold[j]);
		if (del === 0) del = SQRT_EPS;
		const moved = (y[j] + del) - y[j];
		return moved === 0 ? del : moved;
	};
	return {
		dense(t, y, fy, rows) {
			ytry.set(y);
			for (let j = 0; j < neq; j++) {
				const del = increment(y, j);
				ytry[j] = y[j] + del;
				rhs(t, ytry, ftry);
				ytry[j] = y[j];
				for (let i = 0; i < neq; i++) rows[i][j] = (ftry[i] - fy[i]) / del;
			}
			return neq;
		},
		sparse(t, y, fy, pattern, groups, out) {
			const { colPtr, rowIdx } = pattern;
			let spent = 0;
			const perturb = (cols) => {
				ytry.set(y);
				for (const j of cols) {
					delta[j] = increment(y, j);
					ytry[j] = y[j] + delta[j];
				}
				rhs(t, ytry, ftry);
				spent++;
				for (const j of cols) {
					for (let p = colPtr[j]; p < colPtr[j + 1]; p++) {
						out[p] = (ftry[rowIdx[p]] - fy[rowIdx[p]]) / delta[j];
					}
				}
			};
			if (groups) for (const g of groups) perturb(g);
			else for (let j = 0; j < neq; j++) perturb([j]);
			return spent;
		},
	};
}

export const rosenbrockMethod = {
	id: 'ros23',
	order: 2,
	defaultMaxSteps: 1e6,
	holdsAtZero: true,
	snapsToConstraint: true,

	stepBudgetMessage: (maxSteps, t) => `Exceeded ${maxSteps} steps at t=${t}.`,
	floorMessage: (t, hmin) => (
		`Unable to meet integration tolerances at t=${t} without reducing `
		+ `the step below the smallest allowed (${hmin}).`
	),
	stallMessage: (t, window, spanCovered) => (
		`The solver stopped making progress at t=${t}: ${window} steps `
		+ `advanced the clock by less than ${spanCovered}. This usually means a `
		+ 'state reaching zero while its equations push it below -- a constraint '
		+ 'that binds is not one a one-step method can carry; turn "cannot go '
		+ 'negative" off on the compartment to see what the model really does, '
		+ 'or use variableOrder -- or that a rate changes faster than the step size '
		+ 'can follow.'
	),

	/**
	 * `opts.jacobian` supplies df/dy instead of differencing it: a function
	 * (t, y, J) filling dense rows, or the object ../sim/jacobian.js builds,
	 * whose `constant` flag says the matrix never changes -- and since this
	 * method re-forms its Jacobian at every step, that flag turns n+1
	 * evaluations per step into one. An `evaluate` that answers null has no
	 * usable matrix at that point, and the step differences instead.
	 */
	setUp({ neq, rhs, threshold, opts }) {
		const supplied = opts.jacobian ?? null;
		const analytic = typeof supplied === 'function'
			? supplied
			: supplied?.evaluateDense
				? (t, y, J) => supplied.evaluateDense(t, y, J)
				: null;
		const constant = !!supplied?.constant;
		const pattern = supplied?.pattern ?? null;
		const groups = Array.isArray(supplied?.groups) ? supplied.groups : null;
		const diff = differences(rhs, neq, threshold);

		const J = new Array(neq);
		for (let i = 0; i < neq; i++) J[i] = new Float64Array(neq);
		let values = pattern ? new Float64Array(pattern.nnz) : null;
		const f0 = new Float64Array(neq);
		const f1 = new Float64Array(neq);
		const f2 = new Float64Array(neq);
		const ft = new Float64Array(neq);
		const tmp = new Float64Array(neq);
		let k1 = null;
		let k2 = null;
		let npds = 0;
		let ndecomps = 0;
		let spent = 0;
		let needJacobian = true;
		let dfdtAt = NaN; // the start time df/dt was last formed for
		// The step the interpolant belongs to.
		let tFrom = 0;
		let hFrom = 0;
		let yFrom = null;

		/** The pattern's values at (t, y): supplied, or differenced through it. */
		const sparseJacobian = (t, y) => {
			const got = supplied.evaluate(t, y);
			if (got) return got;
			spent += diff.sparse(t, y, f0, pattern, groups, values);
			return values;
		};
		/** Dense rows at (t, y): supplied, or differenced. */
		const denseJacobian = (t, y) => {
			if (analytic && analytic(t, y, J) !== null) return J;
			spent += diff.dense(t, y, f0, J);
			return J;
		};
		/** A pattern's values scattered into dense rows, for a factorisation the fill made dense. */
		const scatter = (vals) => {
			for (let i = 0; i < neq; i++) J[i].fill(0);
			for (let j = 0; j < neq; j++) {
				for (let p = pattern.colPtr[j]; p < pattern.colPtr[j + 1]; p++) J[pattern.rowIdx[p]][j] = vals[p];
			}
		};

		// The linear algebra: sparse where a pattern was given and the fill
		// says it pays, dense otherwise.
		let sparse = null;
		let dense = null;
		let lu = null;
		const solveWith = (h, t) => {
			if (sparse) {
				try {
					sparse.form(h * D, values);
				} catch (e) {
					throw new SolverError(`${e.message} (at t=${t})`, t);
				}
				ndecomps++;
				return (b) => sparse.solve(b);
			}
			for (let i = 0; i < neq; i++) {
				const row = dense[i];
				const src = J[i];
				for (let j = 0; j < neq; j++) row[j] = -h * D * src[j];
				row[i] += 1;
			}
			lu.factorize(dense);
			if (lu.singular) {
				throw new SolverError(
					`The iteration matrix is singular at t=${t}. A compartment may be `
					+ 'disconnected or a transfer rate may be non-finite.', t,
				);
			}
			ndecomps++;
			return (b) => lu.solve(b);
		};

		return {
			start(t, y) {
				rhs(t, y, f0);
				if (pattern) {
					// One factorisation of each ordering decides whether the
					// pattern pays; a matrix this size is often better dense.
					values = sparseJacobian(t, y);
					npds++;
					needJacobian = false;
					sparse = sparseIterationMatrix(neq, pattern, values, { hint: SINGULAR_HINT });
					if (!sparse) scatter(values);
				}
				if (!sparse) {
					dense = new Array(neq);
					for (let i = 0; i < neq; i++) dense[i] = new Float64Array(neq);
					lu = new LU(neq);
				}
				return 1;
			},
			derivativeAtStart() {
				return f0;
			},
			attempt(t, y, h, tnew, ynew, err) {
				const before = spent;
				// A constant Jacobian is formed once; anything else at the
				// start of every step, as the method requires.
				if (needJacobian && !(constant && npds > 0)) {
					if (sparse) values = sparseJacobian(t, y);
					else if (analytic || !pattern) denseJacobian(t, y);
					else scatter(sparseJacobian(t, y));
					npds++;
				}
				needJacobian = false;
				// df/dt by a one-sided difference, once per step: it depends on
				// where the step starts and not on how long it is.
				if (dfdtAt !== t) {
					const dt = Math.sign(h) * Math.min(SQRT_EPS * Math.max(Math.abs(t), Math.abs(t + h)), Math.abs(h));
					rhs(t + dt, y, f1);
					spent++;
					for (let i = 0; i < neq; i++) ft[i] = (f1[i] - f0[i]) / dt;
					dfdtAt = t;
				}
				const solve = solveWith(h, t);

				for (let i = 0; i < neq; i++) tmp[i] = f0[i] + h * D * ft[i];
				k1 = solve(tmp);
				for (let i = 0; i < neq; i++) tmp[i] = y[i] + 0.5 * h * k1[i];
				rhs(t + 0.5 * h, tmp, f1);
				for (let i = 0; i < neq; i++) tmp[i] = f1[i] - k1[i];
				k2 = solve(tmp);
				for (let i = 0; i < neq; i++) k2[i] += k1[i];
				for (let i = 0; i < neq; i++) ynew[i] = y[i] + h * k2[i];
				rhs(tnew, ynew, f2);
				for (let i = 0; i < neq; i++) {
					tmp[i] = f2[i] - E32 * (k2[i] - f1[i]) - 2 * (k1[i] - f0[i]) + h * D * ft[i];
				}
				const k3 = solve(tmp);
				spent += 2;
				for (let i = 0; i < neq; i++) err[i] = k1[i] - 2 * k2[i] + k3[i];
				tFrom = t;
				hFrom = h;
				yFrom = y;
				const fevals = spent - before;
				return { fevals, scale: 1 / 6 };
			},
			denseAt(tq, out) {
				// The quadratic through the two ends of the step with the
				// slope f(t, y) at its start: P(s) = y + s·h·f0 + s²·h·(k2 - f0).
				const s = (tq - tFrom) / hFrom;
				for (let i = 0; i < neq; i++) {
					out[i] = yFrom[i] + s * hFrom * f0[i] + s * s * hFrom * (k2[i] - f0[i]);
				}
				return out;
			},
			accept(tnew, ynew, reprojected) {
				// The next step starts from the state after the projection,
				// so the derivative it starts from has to be read there.
				let used = 0;
				if (reprojected) { rhs(tnew, ynew, f2); used = 1; }
				f0.set(f2);
				needJacobian = true;
				dfdtAt = NaN;
				return used;
			},
			restart(t, y) {
				rhs(t, y, f0);
				needJacobian = true;
				dfdtAt = NaN;
				return 1;
			},
			stats() {
				return {
					npds, ndecomps,
					sparse: !!sparse,
					fill: sparse ? sparse.fill : null,
				};
			},
		};
	},
};

/**
 * @param {(t: number, y: Float64Array, out: Float64Array) => Float64Array} f
 * @param {number[]|Float64Array} tspan  output grid; first and last bound the run
 * @param {Float64Array} y0
 * @param {object} opts  see ./onestep.js, plus `jacobian`
 * @returns {{ t: Float64Array, y: Float64Array[], stats: object }}
 */
export function rosenbrock23(f, tspan, y0, opts = {}) {
	return integrate(rosenbrockMethod, f, tspan, y0, opts);
}

File: kompartment/python/kompartment/engine/solvers/rosenbrock23.py
"""Rosenbrock (2,3): the stiff one-step method (``src/ode/solvers/rosenbrock23.js``).

The modified Rosenbrock triple of Shampine and Reichelt (the method of
MATLAB's ode23s): one iteration matrix I - h*d*J per step, three solves, a
second-order solution with a third-order error estimate and a quadratic
interpolant. It forms the matrix afresh every step, which makes it steady
through a discontinuity.
"""

from __future__ import annotations

import math
from typing import Any, Dict, Optional, Tuple

import numpy as np

from ..jacobian import difference_increment, difference_jacobian
from . import SolverError
from . import kernels
from .matrix import IterationMatrix
from .onestep import integrate

SINGULAR_HINT = 'A compartment with no way in and no way out will do this.'
D = 1 / (2 + math.sqrt(2))
E32 = 6 + math.sqrt(2)
EPS = 2.0 ** -52
SQRT_EPS = math.sqrt(EPS)


class _Stepper:
    def __init__(self, neq: int, rhs: Any, threshold: np.ndarray, opts: Dict[str, Any]) -> None:
        self.neq = neq
        self.rhs = rhs
        self.threshold = threshold
        supplied = opts.get('jacobian')
        self.pattern = supplied.get('pattern') if supplied else None
        self.groups = supplied.get('groups') if supplied else None
        self.evaluate = supplied.get('evaluate') if supplied else None
        self.constant = bool(supplied.get('constant')) if supplied else False
        self.f0 = np.zeros(neq)
        self.ft = np.zeros(neq)
        self.npds = 0
        self.ndecomps = 0
        self.spent = 0
        self.need_jacobian = True
        self.dfdt_at = math.nan
        self.J: Any = None
        self.W: Optional[IterationMatrix] = None
        self._fused = None
        if kernels.available():
            self._fused = (kernels.kernel('ros_first'), kernels.kernel('ros_second'), kernels.kernel('ros_third'))
            self._tmp = np.zeros(neq)

    def _jacobian(self, t: float, y: np.ndarray) -> Any:
        if self.evaluate is not None:
            got = self.evaluate(t, y)
            if got is not None:
                return np.asarray(got, dtype=float)
        if self.pattern is not None:
            self.spent += len(self.groups) if self.groups is not None else self.neq
            return difference_jacobian(self.rhs, t, y, self.f0, self.pattern, self.groups, self.threshold)
        dl = difference_increment(y, self.threshold)
        J = np.zeros((self.neq, self.neq))
        for j in range(self.neq):
            ytry = y.copy()
            ytry[j] = y[j] + dl[j]
            J[:, j] = (np.asarray(self.rhs(t, ytry)) - self.f0) / dl[j]
        self.spent += self.neq
        return J

    def start(self, t: float, y: np.ndarray) -> int:
        self.f0 = np.asarray(self.rhs(t, y), dtype=float)
        self.J = self._jacobian(t, y)
        self.npds += 1
        self.need_jacobian = False
        self.W = IterationMatrix(self.neq, self.pattern, self.J, 'auto', None, {'singular': SINGULAR_HINT})
        return 1

    def derivative_at_start(self) -> np.ndarray:
        return self.f0

    def attempt(self, t: float, y: np.ndarray, h: float, tnew: float, ynew: np.ndarray,
                err: np.ndarray) -> Tuple[int, float]:
        before = self.spent
        if self.need_jacobian and not (self.constant and self.npds > 0):
            self.J = self._jacobian(t, y)
            self.npds += 1
        self.need_jacobian = False
        if self.dfdt_at != t:
            dt = math.copysign(min(SQRT_EPS * max(abs(t), abs(t + h)), abs(h)), h)
            f1 = np.asarray(self.rhs(t + dt, y), dtype=float)
            self.spent += 1
            self.ft = (f1 - self.f0) / dt
            self.dfdt_at = t
        try:
            self.W.form(h * D, self.J)  # type: ignore[union-attr]
        except RuntimeError as e:
            raise SolverError('singular', f'{e} (at t={t})', t) from None
        self.ndecomps += 1
        W = self.W
        if self._fused is not None and W._lu == 'kernel':  # type: ignore[union-attr]
            return self._attempt_fused(t, y, h, tnew, ynew, err, before)
        solve = W.solve  # type: ignore[union-attr]
        f0, ft = self.f0, self.ft
        k1 = solve(f0 + h * D * ft)
        f1 = np.asarray(self.rhs(t + 0.5 * h, y + 0.5 * h * k1), dtype=float)
        k2 = solve(f1 - k1) + k1
        ynew[:] = y + h * k2
        f2 = np.asarray(self.rhs(tnew, ynew), dtype=float)
        k3 = solve(f2 - E32 * (k2 - f1) - 2 * (k1 - f0) + h * D * ft)
        self.spent += 2
        err[:] = k1 - 2 * k2 + k3
        self.t_from, self.h_from, self.y_from = t, h, y.copy()
        self.k2 = k2
        self.f2 = f2
        return self.spent - before, 1 / 6

    def _attempt_fused(self, t: float, y: np.ndarray, h: float, tnew: float, ynew: np.ndarray, err: np.ndarray,
                       before: int) -> Tuple[int, float]:
        """The same step with its arithmetic compiled (``kernels.py``): the
        application's loops, value for value."""
        first, second, third = self._fused  # type: ignore[misc]
        W = self.W
        lu, piv, tmp = W._lu_buf, W._piv, self._tmp  # type: ignore[union-attr]
        n = self.neq
        f0, ft = self.f0, self.ft
        hD = h * D
        k1 = np.empty(n)
        ymid = np.empty(n)
        first(lu, piv, f0, ft, hD, y, 0.5 * h, tmp, k1, ymid)
        f1 = np.asarray(self.rhs(t + 0.5 * h, ymid), dtype=float)
        k2 = np.empty(n)
        second(lu, piv, f1, k1, y, h, tmp, k2, ynew)
        f2 = np.asarray(self.rhs(tnew, ynew), dtype=float)
        k3 = np.empty(n)
        third(lu, piv, f2, k2, f1, k1, f0, ft, hD, E32, tmp, k3, err)
        self.spent += 2
        self.t_from, self.h_from, self.y_from = t, h, y.copy()
        self.k2 = k2
        self.f2 = f2
        return self.spent - before, 1 / 6

    def dense_at(self, tq: float) -> np.ndarray:
        s = (tq - self.t_from) / self.h_from
        return self.y_from + s * self.h_from * self.f0 + s * s * self.h_from * (self.k2 - self.f0)

    def accept(self, tnew: float, ynew: np.ndarray, reprojected: bool) -> int:
        used = 0
        if reprojected:
            self.f2 = np.asarray(self.rhs(tnew, ynew), dtype=float)
            used = 1
        self.f0 = self.f2.copy()
        self.need_jacobian = True
        self.dfdt_at = math.nan
        return used

    def restart(self, t: float, y: np.ndarray) -> int:
        self.f0 = np.asarray(self.rhs(t, y), dtype=float)
        self.need_jacobian = True
        self.dfdt_at = math.nan
        return 1

    def stats(self) -> Dict[str, Any]:
        return {'npds': self.npds, 'ndecomps': self.ndecomps, 'sparse': bool(self.W and self.W.sparse),
                'fill': self.W.info['fill'] if self.W else None}


class _Method:
    id = 'ros23'
    order = 2
    default_max_steps = 1e6
    holds_at_zero = True
    snaps_to_constraint = True

    @staticmethod
    def step_budget_message(max_steps: float, t: float) -> str:
        return f'Exceeded {int(max_steps)} steps at t={t}.'

    @staticmethod
    def floor_message(t: float, hmin: float, worst: int) -> str:
        return (f'Unable to meet integration tolerances at t={t} without reducing the step below the smallest allowed '
                f'({hmin}).')

    @staticmethod
    def stall_message(t: float, window: int, covered: float) -> str:
        return (f'The solver stopped making progress at t={t}: {window} steps advanced the clock by less than '
                f'{covered}. This usually means a state reaching zero while its equations push it below -- a '
                'constraint that binds is not one a one-step method can carry; turn "cannot go negative" off on the '
                'compartment to see what the model really does, or use variableOrder -- or that a rate changes faster '
                'than the step size can follow.')

    @staticmethod
    def set_up(neq: int, rhs: Any, f: Any, threshold: np.ndarray, atol: np.ndarray, rtol: float,
               opts: Dict[str, Any]) -> _Stepper:
        return _Stepper(neq, rhs, threshold, opts)


def rosenbrock23(f: Any, tspan: Any, y0: np.ndarray, opts: Dict[str, Any]) -> Dict[str, Any]:
    return integrate(_Method, f, tspan, y0, opts)
